### PR TITLE
update status cleanup

### PR DESCRIPTION
**Pull request by sears**
# Description

This PR helps reduce improve code complexity in the update status path.  It includes two functional changes:
 - Bypass the multi-version client when checking to see if the operator's FDB server version is correct, and getting an updated connection string.  Without this change, the operator would incorrectly infer the version the FDB cluster was running at.
 - Always use the client library when invoking `status json`.  I added a simple 1-second retry loop, under the assumption that the FDB multi-version client will eventually get lucky and connect to the version of the cluster that is currently available.  (If not, this is a serious FDB bug, and should be addressed there.)

## Type of change

*Please select one of the options below.*

- Bug fix (non-breaking change which fixes an issue)
- Documentation
- Other

## Discussion

*Are there any design details that you would like to discuss further?*

## Testing

`make test` passes.  We should improve upgrade testing, in general.

## Documentation

*Did you update relevant documentation within this repository?*
Added inline comments

*If this change is adding new functionality, do we need to describe it in our user manual?*
No

*If this change is adding or removing subreconcilers, have we updated the core technical design doc to reflect that?*
N/A

*If this change is adding new safety checks or new potential failure modes, have we documented and how to debug potential issues?*
N/A

## Follow-up

*Are there any follow-up issues that we should pursue in the future?*

The mock admin client does not inject errors, so we're not getting meaningful test coverage by invoking it.

*Does this introduce new defaults that we should re-evaluate in the future?*

This switches the default value of useClientLibrary from false to true, and removes support for false.  I added a deprecation warning to the docs.

### CODE REVIEW

File: controllers/controllers.go [hunk 1]
```diff
@@ -42,6 +42,17 @@ const (
 	podSchedulingDelayDuration = 15 * time.Second
 )
 
+// containsAll determines if one map contains all the keys and matching values
+// from another map.
+func containsAll(current map[string]string, desired map[string]string) bool {
```
Comment (johscheuer) [inline controllers/controllers.go:47]: I believe a better place would be the internal package maybe a new `map_helper` file?

File: controllers/update_status.go [hunk 2]
```diff
@@ -333,15 +315,21 @@ func tryConnectionOptions(cluster *fdbv1beta2.FoundationDBCluster, r *Foundation
 			adminClient, clientErr := r.getDatabaseClientProvider().GetAdminClient(cluster, r)
 
 			if clientErr != nil {
+				// If this fails, it means we couldn't instantaite a local object.  GetAdminClient does
+				// not attempt to connect to the FDB server.
```
Comment (johscheuer) [inline controllers/update_status.go:319]:
```suggestion
				// If this fails, it means we couldn't instantiate a local object.  GetAdminClient does
				// not attempt to connect to the FDB server.
```

File: fdbclient/admin_client.go [hunk 3]
```diff
@@ -75,10 +75,6 @@ type cliAdminClient struct {
 	// custom parameters that should be set.
 	knobs []string
 
-	// Whether the admin client should be able to run operations through the
-	// client library rather than the CLI.
-	useClientLibrary bool
```
Comment (johscheuer) [inline fdbclient/admin_client.go:80]: I think in the past @brownleej added this for doing some tests with DNS, but I don't remember the specifics and I'm happy to remove this code if we don't need it anymore.

File: fdbclient/admin_client.go [hunk 4]
```diff
@@ -313,21 +296,24 @@ func (client *cliAdminClient) GetStatus() (*fdbv1beta2.FoundationDBStatus, error
 	adminClientMutex.Lock()
 	defer adminClientMutex.Unlock()
 
-	status, err := client.getStatus(client.useClientLibrary)
-	if err != nil {
-		return nil, err
-	}
+	for {
```
Comment (johscheuer) [inline fdbclient/admin_client.go:299]: Should we limit the number of retries?

File: controllers/admin_client_mock.go [hunk 5]
```diff
@@ -532,6 +532,10 @@ func (client *mockAdminClient) GetConnectionString() (string, error) {
 	return client.Cluster.Status.ConnectionString, nil
 }
 
+func (client *mockAdminClient) GetConnectionStringFromCLI() (string, error) {
```
Comment (johscheuer) [inline controllers/admin_client_mock.go:535]: Do you think we still need this one, once we get https://github.com/FoundationDB/fdb-kubernetes-operator/pull/1421 merged? I think a better approach to using the fdbcli to determine the running version is to use the information from the status that we fetch in the next step in updateStatus.

For reference, this is the place where we check the running version: https://github.com/FoundationDB/fdb-kubernetes-operator/blob/main/controllers/update_status.go#L99-L115 with a simple heuristic (just checking what version the actual fdbserver processes are running).

**Review by johscheuer — CHANGES_REQUESTED** [review]
Could you run `make fmt lint` and fix the resulting issues? That should solve the lint issues.

File: fdbclient/admin_client.go (hunk 3 above)
Comment (sears) [reply fdbclient/admin_client.go:80]: I don't really know of a good way to test this change, but (at least in theory), we can delete it.

Comment (johscheuer) [reply fdbclient/admin_client.go:80]: I think we can drop this feature since it was only used once for some manual testing.

**Pull request closed without merging**